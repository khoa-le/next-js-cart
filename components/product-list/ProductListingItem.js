import styled from "styled-components";
import Link from "next/link";

const ProductListingItemWrap = styled.div`
  color: red;
`;
const PriceBox = styled.div`
  text-decoration: line-through;
`;

const ProductListingItem = ({ product }) => {
  return (
    <ProductListingItemWrap className="col-sm-6 col-lg-4 mb-9">
      <Link href={`/products/${product.product_id}`}>
        <a className="d-block text-dark text-center transition-3d-hover">
          <img
            className="img-fluid rounded"
            src={`/media/product_images/${product.image}`}
            alt={product.name}
          />
          <span className="pt-4 px-4">
            <h2 className="h5 mb-1">{product.name}</h2>
            {product.discounted_price > 0 ? (
              <>
                <PriceBox>{product.price}</PriceBox>
                <span className="d-block">{product.discounted_price}</span>
              </>
            ) : (
              <span className="d-block">{product.price}</span>
            )}
          </span>
        </a>
      </Link>
    </ProductListingItemWrap>
  );
};

export default ProductListingItem;
