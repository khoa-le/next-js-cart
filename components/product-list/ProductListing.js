import styled from "styled-components";
import { GlobalConfiguration } from "../../utils/global";
import ProductListingItem from "./ProductListingItem";
import Pagination from "./Pagination";

const ProductListingWrapper = styled.div``;

const ProductListing = ({ products }) => {
  return (
    <div className="container">
      <ProductListingWrapper className="row mx-gutters-5">
        {products.length == 0 ? (
          <p>Product listing was empty</p>
        ) : (
          <>
            {Object.keys(products).map((key) => {
              if (key < GlobalConfiguration.ProductListing.PerPage) {
                const product = products[key];
                return (
                  <ProductListingItem
                    product={product}
                    key={product.product_id}
                  />
                );
              }
            })}
          </>
        )}
      </ProductListingWrapper>
      <Pagination numberProducts={products.length}></Pagination>
    </div>
  );
};
export default ProductListing;
