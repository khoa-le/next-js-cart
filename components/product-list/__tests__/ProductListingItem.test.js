import ProductListingItem from "../ProductListingItem";
import { render, cleanup } from "@testing-library/react";
import "jest-styled-components";
import renderer from "react-test-renderer";
import styled from "styled-components";
import ProductListing from "../ProductListing";

const product = {
  product_id: 1,
  image: "12312321",
  name: "Khoa Product",
  price: 4321,
  discounted_price: 0,
};

afterEach(cleanup);

describe("Product Listing Item", () => {
  test("Can render Product Listing Item", () => {
    const { getByText } = render(<ProductListingItem product={product} />);
    const linkElement = getByText("Khoa Product");
    expect(linkElement).toBeInTheDocument();
  });

  test("See discount price if they have", () => {
    product.discounted_price = 1234;
    const { getByText } = render(<ProductListingItem product={product} />);
    const discountedPriceElement = getByText("1234");
    expect(discountedPriceElement).toBeInTheDocument();
  });
  test("See discount price if they have", () => {
    product.discounted_price = 0;
    const { getByText } = render(<ProductListingItem product={product} />);
    const priceElement = getByText("4321");
    expect(priceElement).toBeInTheDocument();
  });
  test("See current price is Line-through when product have discount", () => {
    product.discounted_price = 1234;
    const { container, getByText } = render(
      <ProductListingItem product={product} />
    );
    expect(container.firstChild).toMatchSnapshot();
    const priceBoxElement = getByText("4321");
    expect(priceBoxElement).toHaveStyleRule("text-decoration", "line-through");
  });
});
