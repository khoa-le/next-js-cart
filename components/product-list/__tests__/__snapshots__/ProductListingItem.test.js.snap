// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Listing Item See current price is Line-through when product have discount 1`] = `
.c0 {
  color: red;
}

.c1 {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

<div
  class="c0 col-sm-6 col-lg-4 mb-9"
>
  <a
    class="d-block text-dark text-center transition-3d-hover"
    href="/products/1"
  >
    <img
      alt="Khoa Product"
      class="img-fluid rounded"
      src="/media/product_images/12312321"
    />
    <span
      class="pt-4 px-4"
    >
      <h2
        class="h5 mb-1"
      >
        Khoa Product
      </h2>
      <div
        class="c1"
      >
        4321
      </div>
      <span
        class="d-block"
      >
        1234
      </span>
    </span>
  </a>
</div>
`;
